tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/stratifiedRules/aux/caches/UninterpretedLiteralCache.scala

```diff
@@ -0,0 +1,74 @@
+package at.forsyte.apalache.tla.bmcmt.stratifiedRules.aux.caches
+
+import at.forsyte.apalache.tla.bmcmt.smt.SolverContext
+import at.forsyte.apalache.tla.bmcmt.types.{CellT, CellTFrom}
+import at.forsyte.apalache.tla.bmcmt.{ArenaCell, PureArena}
+import at.forsyte.apalache.tla.lir.{ConstT1, StrT1}
+import at.forsyte.apalache.tla.types.tla
+
+/**
+ * A cache for uninterpreted literals, that are translated as uninterpreted SMT constants, with a unique sort per
+ * uninterpreted type. Since two values are equal iff they are literally the same literal, we force inequality between
+ * all the respective SMT constants.
+ *
+ * Note that Strings are just a special kind of uninterpreted type.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+class UninterpretedLiteralCache extends Cache[PureArena, (String, String), ArenaCell] {
+
+  /**
+   * Given a pair `(utype,idx)`, where `utype` represents an uninterpreted type name (possibly "Str") and `idx` some
+   * unique index within that type, returns an extension of `arena`, containing a cell, which represents "idx_OF_utype"
+   * (or "idx", if utype = "Str"), and said cell.
+   *
+   * Note that two values are equal (and get cached to the same cell) iff they have the same type and the same index, so
+   * e.g. "1_OF_A" and "1_OF_B" (passed here as ("A", "1") and ("B", "1")) get cached to different, incomparable cells,
+   * despite having the same index "1".
+   */
+  protected def create(
+      arena: PureArena,
+      typeAndIndex: (String, String)): (PureArena, ArenaCell) = {
+    val (utype, _) = typeAndIndex
+    // introduce a new cell
+    val cellType = if (utype == StrT1.toString) StrT1 else ConstT1(utype)
+    val newArena = arena.appendCell(CellT.fromType1(cellType))
+    (newArena, newArena.topCell)
+  }
+
+  /**
+   * The UninterpretedLiteralCache maintains that a cell cache for a value `idx` of type `tp` is distinct from all other
+   * values of type `tp` (defined so far).
+   *
+   * Whenever possible, try to use [[addAllConstraints]] instead of this method, for performance reasons instead.
+   */
+  override def addConstraintsForElem(ctx: SolverContext): (((String, String), ArenaCell)) => Unit = {
+    case ((tp, _), v) =>
+      val cellType = if (tp == StrT1.toString) StrT1 else ConstT1(tp)
+      val others = values().withFilter { c => c.cellType == CellTFrom(cellType) && c != v }.map(_.toBuilder).toSeq
+      // The fresh cell should differ from the previously created cells.
+      // We use the SMT constraint (distinct ...).
```

Comment:
Why don't we need a check that the entry is actually in the cache?

Reply:
We could add it, but there's technically nothing wrong is asserting that a value outside the cache is distinct from all values in the cache. Theoretically, once this is wired to some manager class, this method will either:
 - get called once, immediately after an element is added, so membership is established by construction, or
 - never get called in favor of `addAll`, so the issue is irrelevant

it's just that, _at this point in time_, I cannot 100% predict what the integration of this class into the rules engine will look like, since it won't exactly be 1-to-1 with `ModelValueCache`. If you want a QOL exception trigger, I'd be happy to put that in though.